src/extension.ts

```diff
@@ -1588,6 +1604,66 @@ export async function activate(context: vscode.ExtensionContext): Promise<any> {
   return extensionApi;
 }
 
+// This function is exported as one of our API functions but is also used internally
+// for example to implement the async variant capable of resolving docker port number.
+function serverForUri(uri: vscode.Uri): any {
+  const { apiTarget } = connectionTarget(uri);
+  const api = new AtelierAPI(apiTarget);
+
+  // This function intentionally no longer exposes the password for a named server UNLESS it is already exposed as plaintext in settings.
+  // API client extensions should use Server Manager 3's authentication provider to request a missing password themselves,
+  // which will require explicit user consent to divulge the password to the requesting extension.
+  const {
+    serverName,
+    active,
+    host = "",
+    https,
+    port,
+    pathPrefix,
+    username,
+    password,
+    ns = "",
+    apiVersion,
+    serverVersion,
+  } = api.config;
+  return {
+    serverName,
+    active,
+    scheme: https ? "https" : "http",
+    host,
+    port,
+    pathPrefix,
+    username,
+    password:
+      serverName === ""
+        ? password
+        : vscode.workspace.getConfiguration(`intersystems.servers.${serverName.toLowerCase()}`, uri).get("password"),
+    namespace: ns,
+    apiVersion: active ? apiVersion : undefined,
+    serverVersion: active ? serverVersion : undefined,
+  };
+}
+
+// An async variant capable of resolving docker port number.
+// It is exported as one of our API functions but is also used internally.
+async function asyncServerForUri(uri: vscode.Uri): Promise<any> {
```

Comment:
What does this mean for Language Server? Does it have to change to support these `docker compose` connections?

Reply:
I don't think so. It will continue to ask the ObjectScript extension for the connection details (host, port, pathPrefix etc) as usual. This extension does the magic.